### PR TITLE
ScreenSaver: Cache image paths for random image

### PR DESCRIPTION
**Pull request by melyux**
Coming in with a more original addition. For a while, I had a dream to put the entire history of a comic strip into KOReader's "screenshots" folder and show a random strip on every lock. This was blocked by ~~two things~~ one thing:

~~1) I'd have to put all the images in the root folder of screenshots instead of having them stay organized in their folders~~
2) It is incredibly, incredibly slow

These are my working solution to ~~these problems~~ this problem:

~~1) I allow `Screensaver:_getRandomImage()` to traverse the screenshots folder recursively to grab all images from all nested subdirectories.~~
2) After the first traversal, I create a hidden cache file in the screenshots folder that caches all the image paths found and stores them there for quick random path grabbing thereafter, with the cache effortlessly and automatically becoming invalid when the user adds or deletes or moves any image files in the directory, without any extra software or user interaction or tracking required.

With this change, I can simply drop my ~3600 comic strips ~~(which were already organized into various subfolders)~~ right into the screenshots folder. After the first screensaver, using a random image becomes **instant** instead of taking several seconds.

Some benchmarks:
- Before this change, if I drop 3600 images into the root folder of screenshots and use random image, it takes KOReader **16 seconds** _every single suspend_ to choose a random image. ~~If the images are not directly in the screenshots folder's root, KOReader does not find the them.~~
- After this change, if I put 3600 images into the root folder, ~~organized into any number of subdirectories and sub-subdirectories and so on that I like,~~ and it takes KOReader **18 seconds** on the _first_ suspend to choose a random image and create a cache, and then takes only a **fraction of a second** on every suspend thereafter to choose a random image. Even with thousands of files, I didn't find any difference in delay between 1 image and 5000 images. The cache is fast.

When the user adds a new screensaver image, the cache is automatically rebuilt on the next suspend by comparing the directory modify timestamp to the one recorded in the cache file. If there is no cache file, one is built. If the cache file is invalid, a new one is built. The cache is decentralized, the cache is self-invalidating, the cache is hands-off.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9910)

### CODE REVIEW

**Comment by melyux** [comment]
Ok, I lied. One problem is that the directory modify time can't detect changes in subdirectories, so my two changes clash with one another. If we didn't add recursion, the directory modtime would work. Looking into it right now

**Comment by NiLuJe** [comment]
I'm not necessarily a fan of making this recursive (at least by default), FWIW.

Also: why the custom on-disk format instead of just using Persist to store a ready-to-use table/array or tables?

**Comment by melyux** [comment]
@NiLuJe Because this is the first time I'm hearing of Persist :)

**Comment by NiLuJe** [comment]
Note to self: LuaDoc that thing up ;).

**Comment by melyux** [comment]
Gave up on the recursion, now back to being root-folder only. Luckily, you can select your screensaver folder, so it's not a huge deal (besides the file browser being extremely slow to render even the parent directory of your subdirectory containing 3600 images so you can actually choose the folder as your screensaver folder...)

**Comment by melyux** [comment]
Hmm, so the cache never get re-built no matter when you change files?

**Comment by Frenzie** [comment]
Note that there are some minor inefficiencies in the code that become more relevant once you exceed a thousand files or so. I've detailed them… somewhere. But there's always the I/O factor regardless.

**Comment by poire-z** [comment]
> Hmm, so your cache never gets re-built even if you change files?

What cache ? Yours from this PR ? Didn't try it.
Other caches from KOReader ? Dunno if any use that kind of directory modification time dependancy.

**Comment by melyux** [comment]
I don't mind leaving this in my private patches collection. It works great on my Kindle but obviously depends on its stack, which may not be universal. I don't want to add device-specific code that has to be maintained either, but wanted to share it since the speed gains are considerable

**Comment by Frenzie** [comment]
> Other caches from KOReader ? Dunno if any use that kind of directory modification time dependancy.

The concept could potentially make a lot of sense in combination with persist for regular file browser use. Well, if the mtime is actually a tiny bit dependable, lol.

**Comment by poire-z** [comment]
> Do try it, I need some proof of death before I give up the ghost on this one...

Won't be able to, until later this evening.

But I just tried the same thing on a FAT32 USB stick on Wiindows, and the dir mod time is also not udpated.
But on a NTFS filesystem, it is.

What's your OS ? MacOSX ?
What's the Kindle exported filesystem type ?

**Comment by NiLuJe** [comment]
> What's the Kindle exported filesystem type ?

FAT32

**Comment by NiLuJe** [comment]
> Anyway, just tried to copy a file (created at 10h04, copied at 10h04) from Windows over USB to my Kobo, and the target directory modification time didn't change:

Was that a *new* file, or a modification of an existing one?

The former *should* update the directory's mtime (provided it's a direct descendant of said directory, and not in a subdirectory), but *not* the latter... At least on POSIX-compliant systems/fs.

**Comment by Frenzie** [comment]
> At least on POSIX-compliant systems/fs.

Which famously describes FAT to a t, lol

**Comment by NiLuJe** [comment]
This likely smells like a Windows "optimization", because it behaves fine when the switcheroo is done natively on a vfat mount:

```
┌─(ROOT@dahlia:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(0.39:17%:18:43:36%:#)── mkdir /mnt/onboard/foo                                                                                                                                                                                                                                                                                                      ──(Tue, Dec 13)─┘
- correct /mnt/onboard/foo to /mnt/onboard/.foo ? ([Y]es/[N]o/[E]dit/[A]bort) -n
┌─(ROOT@dahlia:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(0.33:17%:18:43:36%:#)── stat /mnt/onboard/foo                                                                                                                                                                                                                                                                                                       ──(Tue, Dec 13)─┘
  File: /mnt/onboard/foo
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: b303h/45827d    Inode: 9552        Links: 2
Access: (0755/drwxr-xr-x)  Uid: (    0/   admin)   Gid: (    0/    root)
Access: 2022-12-13 18:43:34.000000000 +0100
Modify: 2022-12-13 18:43:34.000000000 +0100
Change: 2022-12-13 18:43:34.000000000 +0100
┌─(ROOT@dahlia:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(0.30:17%:18:43:36%:#)── touch /mnt/onboard/foo/bar                                                                                                                                                                                                                                                                                                  ──(Tue, Dec 13)─┘
┌─(ROOT@dahlia:pts/0)──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~)─┐
└─(0.28:17%:18:43:36%:#)── stat /mnt/onboard/foo                                                                                                                                                                                                                                                                                                       ──(Tue, Dec 13)─┘
  File: /mnt/onboard/foo
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: b303h/45827d    Inode: 9552        Links: 2
Access: (0755/drwxr-xr-x)  Uid: (    0/   admin)   Gid: (    0/    root)
Access: 2022-12-13 18:43:42.000000000 +0100
Modify: 2022-12-13 18:43:42.000000000 +0100
Change: 2022-12-13 18:43:42.000000000 +0100
```

**Comment by NiLuJe** [comment]
Also, because I stumbled upon it while attempting to see if this was documented behavior:

https://learn.microsoft.com/en-us/windows/win32/sysinfo/file-times?redirectedfrom=MSDN

TL;DR: This is why https://github.com/koreader/koreader/commit/1ca0748ca74b3daca9b1847776da6f144edaf3bb is not so crazy after all: timestamps on FAT32 are fucked by design ;).

**Comment by poire-z** [comment]
> Was that a _new_ file, or a modification of an existing one?
> 
> The former _should_ update the directory's mtime

It was this former, and it did not.
I'd like somebody else to verify that behaviour on his own Windows 10/11, using file explorer to copy/paste books/files.

**Comment by poire-z** [comment]
Had a try with this PR on my Kobo.
Created empty screensaver/ directory at 19h48
Put one file in it at 19h48.
USB unplug, koreader launched, suspend, file is shown.

USB replug:
```
$ la ../screensaver
drwxr-xr-x 1 0 Dec 13 19:48 .
-rw-r--r-- 1 112 Dec 13 19:48 .screensaverCache
-rw-r--r-- 1 104918 Dec 13  2005 file1.jpg
```
Added a 2nd file at 19h50,
USB unplug, koreader launched, suspend multiple times, only first file is shown.

USB replug:
```
drwxr-xr-x 1 0 Dec 13 19:48 .
-rw-r--r-- 1 211 Dec 13 19:50 .screensaverCache
-rw-r--r-- 1 104918 Dec 13  2005 file1.jpg
-rw-r--r-- 1 120872 May 25  2006 file2.jpg
```
The .screensaverCache has been updated to include the 2nd file !  But still first file shown.
Added a 3nd file at 19h56, added some log in screensaver.lua

USB unplug, koreader launched, suspend three times, still only first file is shown.
Went over telnet do the `ls` :

```
/mnt/onboard/.adds/screensaver # ls -la
drwxr-xr-x    2 root     root          8192 Dec 13 19:48 .
drwxr-xr-x   21 root     root          8192 Dec 11 00:58 ..
-rwxr-xr-x    1 root     root           211 Dec 13 19:50 .screensaverCache
-rwxr-xr-x    1 root     root         25375 Dec 13  2005 file3.jpg
-rwxr-xr-x    1 root     root        104918 Dec 13  2005 file1.jpg
-rwxr-xr-x    1 root     root        120872 May 25  2006 file2.jpg
```
This time, the cache file has not been updated.
From the added log:
```
12/13/22-19:57:14 WARN  tscache 1670957330 tsreal 1670957330 true
12/13/22-19:57:20 WARN  tscache 1670957330 tsreal 1670957330 true
12/13/22-19:57:24 WARN  tscache 1670957330 tsreal 1670957330 true
```

1670957330 = 19h48

So, well, may be a bug when starting with 1 file - but still the timestamps are not updated (for me, on my Windows, with my Kobo, under KSM).
Content of the `.screensaverCache` (padding removed):
```
1670957330
2
file1.jpg
file2.jpg
```

**Comment by melyux** [comment]
I tried starting with one file and building up, but it worked for me. So likely entirely due to your directory mtime not updating when files are added or removed. That seems... like a broken af filesystem. What else is the point of directory mtimes if they're not gonna update on actual changes

**Comment by Frenzie** [comment]
@poire-z Do you mean on Windows 10/11 with FAT32/VFAT in general or with a specific procedure connecting to the Kobo?

I ask because I don't really trust Windows near USB FAT filesystems. :-P

**Comment by NiLuJe** [comment]
I would hope it behaves the same on all FAT32 partitions, although there might be extra trickery involved if they're removable...

Which reminds me of the "safely eject" thingy that has a behavioral switch... somewhere, that got changed a few FU ago to default to do... something that is supposed to prevent FS corruptions on dirty ejects... Not updating metadata sounds like the sort of workaround that would involve ;).

**Comment by poire-z** [comment]
> Do you mean on Windows 10/11 with FAT32/VFAT in general or with a specific procedure connecting to the Kobo?

Yes, tested with removable FAT32 (Kobo, regular USB stick), on 2 Windows 10, always "safely eject" via the system tray icon. (I don't have any non-removable FAT32 systems around :)
No specific procedure: just drop a file in an already created directory there, and whether or not that directory get its date modified.
(On linux fs, or on Windows NTFS, the directory date is modified.)

**Comment by poire-z** [comment]
Closing. Reopen if other ideas or Windows/FAT32 tests showing that my setup is peculiar and uncommon :)

**Comment by NiLuJe** [comment]
Forgot to test on my end when I rebooted last night -_-".

I'll try to remember the next time I boot the Win10 part ;).

**Comment by Frenzie** [comment]
Same, I'd booted into Windows last night and forgot about it, d'oh.

**Comment by NiLuJe** [comment]
Yup, confirmed. On the upside, Windows itself agrees ;).

Kobo (before/during/after USBMS):

```

[admin@dahlia root]# stat /mnt/onboard/.foo/
  File: /mnt/onboard/.foo/
  Size: 16384           Blocks: 32         IO Block: 4096   directory
Device: b303h/45827d    Inode: 5390        Links: 3
Access: (0755/drwxr-xr-x)  Uid: (    0/   admin)   Gid: (    0/    root)
Access: 2022-05-28 01:00:00.000000000 +0200
Modify: 2022-05-28 16:23:46.000000000 +0200
Change: 2020-10-23 20:31:56.000000000 +0200
[admin@dahlia root]# stat /mnt/onboard/.foo/
stat: can't stat '/mnt/onboard/.foo/': No such file or directory
[admin@dahlia root]# stat /mnt/onboard/.foo/
  File: /mnt/onboard/.foo/
  Size: 16384           Blocks: 32         IO Block: 4096   directory
Device: b303h/45827d    Inode: 12567       Links: 3
Access: (0755/drwxr-xr-x)  Uid: (    0/   admin)   Gid: (    0/    root)
Access: 2022-05-28 01:00:00.000000000 +0200
Modify: 2022-05-28 16:23:46.000000000 +0200
Change: 2020-10-23 20:31:56.000000000 +0200
```

(i.e., no changes)

Windows (before/during|after)
```
PS C:\Users\ninuj> Get-Item E:\.foo | Format-List

    Directory: E:\

Name           : .foo
CreationTime   : 23/10/2020 20:31:56
LastWriteTime  : 28/05/2022 16:23:46
LastAccessTime : 28/05/2022 01:00:00
Mode           : d----
LinkType       :
Target         :


PS C:\Users\ninuj> Get-Item E:\.foo | Format-List

    Directory: E:\

Name           : .foo
CreationTime   : 23/10/2020 20:31:56
LastWriteTime  : 28/05/2022 16:23:46
LastAccessTime : 28/05/2022 01:00:00
Mode           : d----
LinkType       :
Target         :


PS C:\Users\ninuj> Get-Item E:\.foo\foo.txt | Format-List

    Directory: E:\.foo

Name           : foo.txt
Length         : 0
CreationTime   : 17/12/2022 20:33:05
LastWriteTime  : 17/12/2022 20:33:06
LastAccessTime : 17/12/2022 00:00:00
Mode           : -a---
LinkType       :
Target         :
VersionInfo    : File:             E:\.foo\foo.txt
                 InternalName:
                 OriginalFilename:
                 FileVersion:
                 FileDescription:
                 Product:
                 ProductVersion:
                 Debug:            False
                 Patched:          False
                 PreRelease:       False
                 PrivateBuild:     False
                 SpecialBuild:     False
                 Language:
```

**Comment by Frenzie** [comment]
Well, that saves me some effort. I'd just pulled out a Mesozoic 2 GB microSD card for testing purposes.

**Comment by NiLuJe** [comment]
That's on a Windows 10 22H2 host (installed in september 2021), FWIW.
